core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java

```diff
@@ -0,0 +1,424 @@
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+
+            if (canceled) {
+                return;
+            }
+            canceled = true;
+
+            completionFuture.completeExceptionally(cause);
+            cancelChildSubscribersAndBuffer();
+            downstream.onError(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            if (canceled) {
+                return;
+            }
+
+            if (canComplete()) {
+                downstream.onComplete();
+                completionFuture.complete(null);
+            } else {
+                completing = true;
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                onError(new IllegalArgumentException(
+                        "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+                cancel();
+                return;
+            }
+
+            if (canceled) {
+                return;
+            }
+
+            if (executor.inEventLoop()) {
+                handleRequest(n);
+            } else {
+                executor.execute(() -> handleRequest(n));
+            }
+        }
+
+        private void handleRequest(long n) {
+            requestedByDownstream = LongMath.saturatedAdd(requestedByDownstream, n);
+            flush();
+
+            if (!initialized) {
+                initialized = true;
+                upstream.request(maxConcurrency);
+            }
+
+            requestAllAvailable();
+        }
+
+        @Override
+        public void cancel() {
+            if (executor.inEventLoop()) {
+                cancel0();
+            } else {
+                executor.execute(() -> cancel0());
+            }
+        }
+
+        private void cancel0() {
+            if (canceled) {
+                return;
+            }
+
+            canceled = true;
+            upstream.cancel();
+            cancelChildSubscribersAndBuffer();
+        }
+
+        private void cancelChildSubscribersAndBuffer() {
+            buffer.forEach(StreamMessageUtil::closeOrAbort);
+            buffer.clear();
+            childSubscribers.forEach(FlatMapSubscriber::cancel);
+        }
+
+        void childSubscribed(FlatMapSubscriber<T, U> child) {
+            pendingSubscriptions--;
+            childSubscribers.add(child);
+
+            requestAllAvailable();
+        }
+
+        private long getAvailableBufferSpace() {
+            if (requestedByDownstream == Long.MAX_VALUE) {
+                return Long.MAX_VALUE;
+            }
+
+            final Optional<Long> requested = childSubscribers.stream().map(FlatMapSubscriber::getRequested)
+                                                             .reduce(LongMath::saturatedAdd);
+
+            return maxConcurrency - requested.orElse(0L) - buffer.size();
```

Comment:
I think you're right, but I'm trying to strike a balance between how much buffering is done between the requests. In the example you gave child2 and child3 will be idle, but at the same time there will be two values in the buffer. Once the downstream requests those 2 values all the children will become active again since the buffer is empty.

If each child subscriber has a fixed target the buffer size will become a lot larger, e.g if `maxConcurrency == 1000`, `requestedByDownstream == 1000` the buffer could contain a million values.